### DISCUSSION TITLE
BIN_IVF_FLAT query slow

### DISCUSSION THREAD

**Discussion by znn0301** [milvus-io/milvus#4280] [Q&A and General discussion] [answered]
records: 410000
indexType: BIN_IVF_FLAT
metricType: HAMMING
nlist:1024
nprobe: 32
topk: 20
milvus.yaml: default, not modified
CPU: 16G
The query result averages 100 milliseconds, is it normal?

**Comment by shiyu22** [comment]
What is your vector dimension? What is the nq(vector num of a query)?

100 milliseconds seems not very fast, so please make sure that other programs are not hogging machine resources when running Milvus.

**Comment by znn0301** [comment]
dimension = 256  nq = 20
The test machine only runs milvus.

**Comment by znn0301** [comment]


**Reply by shiyu22** [reply]
I see, actually binary data has little advantage under IVF_FLAT index, so I think you can try `FLAT` to search, and please let me know if you have the results.

By the way, the FLAT search is related to CPU hardware, how many cores does your machine have?

**Comment by znn0301** [comment]
Number of cores of my machine:
![image](https://user-images.githubusercontent.com/31366133/100304278-54ad8a00-2fd9-11eb-80b6-f4a2bffbd8b7.png)
But isn't FLAT an exact match query? Can't approximate query? 
Sorry, I confirmed again, my index was not created successfully, rebuild the index and ran it again, which took about 80ms. It seems not too fast, and it takes a little long to create the index, about 39 seconds.
The query time is as follows:
top3: 
69、68、67、78、67、77、68、70、68、67、74、75、69、68、70、70、68、67、69、69、66、68、66、69、66、68、68、68、73、68、67、66、67、67、68、67、69、68、68、66、67、67、69、68、70、68、67、68、68、77、79、74、76、70、67、70、68、70、72、73、68、67、68、68、68、69、68、67、67、66、67、69、70、73、70、67、69、68、71、76、77、75、76、69、78、68、67、68、68、69、67、68、69、69、68、68、74、70、69、68、69、70、76、75、76、66、67、78、76、78、67、77、78、67、68、80、68、67、68、70、75、75、67、79、74、77、81、77、80、70、77、67、67、69、70、68、68、67、68、72、68、68、66、68、69、69、68、73、69、71、69、68、69、70、77、81、78、78、76、68、68、68、68、68、67、68、69、67、67、68、67、69、68、67、69、68、68、68、70、72、70、69、67、68、68、71、68、69、68、67、68、69、78、68、71、69、77、68、65、68、79、67、367、67、77、77、78、66、68、67、67、69、77、73、67、66、68、69、68、73、69、67、67、70、69、67、69、68、67、66、68、69、67、67、68、68、70、70、68、67、66、67、68、68、67、68、68、68、66、67、68、70、71、70、68、68、68、68、67、68、68、68、68、67、68、68、67、68、69、69、67、69、69、67、68、68、68、67、68、69、69、67、68、71、80、70、69、74、71、73、80、76、79、80、78、79、67、67、68、75、75、68、70、68、69、69、70、67、68、67、68、68、68
top20:
79、78、78、77、76、78、78、76、79、78、77、76、78、76、76、79、77、78、76、76、78、76、76、77、77、80、81、83、79、80、82、85、79、81、77、82、82、85、77、79、80、80、74、75、78、77、77、79、79、77、77、76、76、77、78、78、77、78、78、78、77、76、78、75、76、77、77、77、80、85、77、78、76、85、85、74、89、83、82、78、89、76、80、80、79、79、77、77、82、86、87、84、87、89、85、81、85、77、76、80、88、80、81、76、77、82、78、89、91、87、79、79、79、79、86、89、78、87、78、87、86、84、75、75、76、77、74、78、75、80、83、88、92、95、91、88、87、88、87、97、87、89、84、89、83、84、79、79、87、89、89、84、81、89、88、82、81、79、81、76、85、78、78、77、78、80、80、76、76、78、74、74、74、76、81、79、82、78、81、81、80、79、86、77、77、77、78、76、80、77、77、75、80、81、82、84、78、76、76、77、78、76、85、86、83、88、76、71、75、76、73、78、75、75、77、80、83、88、85、85、86、87、81、83、87、82、85、86、80、76、76、76、86、75、76、74、77、77、86、76、84、76、76、76、75、80、76、73、76、76、76、76、75、74、79、75、75、77、76、77、76、76、77、76、77、75、75、75、76、76、78、88、78、77、76、77、77、77、77、76、77、85、76、75、76、74、74、76、76、75、76、77、77、74、76、76、76、76、75、77、74、74、78、78、74、75、76、76、75、75、77、76、75、75、76、77、77、76、76、73、76、77、76、76、78、77、79、74、74、75、76、76、75、77、75、76、77、76、75、74、77、76、74、78、75、77、77、80、74、75、75、76、74、74、73、79、77、75、76、74、77、76、77、78、74、77、74、75、75、75、79、78、75、78、76、78、76、75、75、76、76、75、74、76、79、96、79、82、90、85、91、85、80、76、77、76、73、77、82、88、76、76、73、76、73、75、73、74、77、76、77、76、75、84、78、75、75、79、76、75、76、75、75、75、78、75、75、77、78、75、76、75、75、77、74、77、77、75、76、75、75、76、75、76、77、79、79、76、78、75、77、76、78、76、76、82、79、82、80、77、80、78、80、77、85、78、80、77、74、76、83、76、78、78、76、76、78、76、79、77、77、83、79、77、77、78、76、78、81、75、79、75、76、82、75、82、81、76、74、75、76、77、86、75、78、78、76、77、79、77、76、86、81、74、76、77、78、77、76、76、75、76、82、83、81、86、85、82、86、84、85、74、77、76、76、78、80、76、76、75、80、77、76、77、89、83、86、80、81、75、81、88、88、85、89、85、79、78、86、78、78、84、78、74、77、75、77、85、77、87、83、86、91、87、85、88、89、86、85、88、86、82、78、79、78、80、78、77、77、85、84、87、78、76、76、78、78、77、77、77、76、78、76、81、89、86、84、77、76、78、80、78、74、78、87、85、77、75、78、77、78、80、80、77、79、77、76、77、78、77、76、74、76、76、81、87、84、77、74、76、76、75、76、80、74、84、77、77、76、78、79、76、78、79、77、75、73、78、76、75、76、76、79、75、77、86、77、74、84、76、76、78、74、76、75、76、74、77、78、90、88、84、84、86、76、77、80、84、83、86、86、81、78、71、78、77、78、74、77、80、76、76、76、77、76、74、79、72、87、87、85、78、86、75、77、77、76、75、75、76、75、75、75、74、75、76、73、79、74、75、77、72、75、73、74、72、72、74、75、85、75、75、75、80、75、78、78、75、77、76、76、80、77、87、82、83、76、77、76、79、75、83、80、76、75、75、86、74、79、78、78、76、76、84、85、76、84、85、85、85、84、79、74、74、78、74、77、79、77、85、78、78、76、77、74、73、75、73、75、77、74、75、74、74、76、74、74、76、74、73、74、77、73、75、75、73、74、74、74、75、77、75

**Comment by shiyu22** [comment, marked as answer]
Your machine looks good. Which version of Milvus do you use?

You can also refer to it for [performance tuning](https://milvus.io/docs/v0.11.0/tuning.md).

FLAT is indeed an exact search, but do you want to use an approximate search instead of FLAT? In fact, we recommend using FLAT for Binary data.

**Comment by znn0301** [comment]
Milvus version : 0.11.0 ; java-sdk: 0.9.1
Yes, it needs approximate search. What we store here is the hash value of the picture, and the similarity of the picture is obtained by calculating the Hamming distance.
I would also like to ask whether it supports the calculation of similarity with 100-dimensional 32-bit integers. It is also through Hamming distance. A 32-bit integer is equivalent to a bit. Find the number of different integers to get the distance.

**Reply by shiyu22** [reply]
In fact, exact calculations will also return vector similarities.

If you have 100-dimensional integer data, you can convert it to 800-dimensional binary data and calculate the Hamming distance.
This means that Milvus can only calculate the Hamming distance for binary data, so if you need to calculate the Hamming distance for int data, please submit us a [feature issue](https://github.com/milvus-io/milvus/issues).
